src/peer_select.cc

```diff
@@ -530,7 +530,7 @@ PeerSelector::noteIp(const Ip::Address &ip)
 
     Comm::ConnectionPointer p = new Comm::Connection();
     p->remote = ip;
-    p->remote.port(peer ? peer->http_port : request->url.port());
+    p->remote.port(peer ? peer->http_port : request->url.port().value_or(0));
```

Comment:
This `.value_or(0)` is broken. The remote port should never be _unknown_ at this point (though it may be `0` for localhost traffic).
If it is unset/unknown something has gone badly wrong with forwarding and the peer/server cannot be selected.

Please either fix (throw an error ?) , or add XXX to document the bug.

Reply:
These Connection are intended to be (if not actually) opened. That requires a valid port number before it is sent to `handlePath()` below. The code leading up to peer selection is responsible for ensuring that the request-target is at least valid as a destination. Otherwise it is not worth even doing the IP lookup on the non-peer destinations.

(IP cache is a "red herring", this is the caller and obviously has access to the `request->url` destination details).

Reply:
> Otherwise it is not worth even doing the IP lookup on the non-peer destinations.

In short, I believe there are two reasonable solutions here:

1. Change PeerSelector so that it never selects DIRECT destinations for request-targets with unknown ports, under the assumption that Squid does not have any use cases and does not support any protocols that would be able to use a portless DIRECT destination. This forwarding route selection logic change is clearly outside this PR scope, even if it does not result in any functionality changes for existing Squid installations.

2. Continue to let the caller decide what DIRECT selections are valid (for caller purposes). This is what the current PR code is already doing.

> > Please either fix (throw an error ?) , or add XXX to document the bug.

Added an XXX in commit 41334f0. Please unblock this PR.

Reply:
Dislike, but will stop objecting.

Reply:
Oh, as for out of scope: this is the non-tunnel equivalent piece of the forwarding bug half-fixed in this PR with reasoning https://github.com/squid-cache/squid/pull/1255/files/faeb96979ae6cfb0829a2676cb9751efeed48996#r1095191336

Reply:
> this is the non-tunnel equivalent

The critical difference between the two areas is that the high-level ConnStateData::initiateTunneledRequest() code already had logic for handling missing information -- that whole method is essentially about those missing info checks, and there was no question there that a port is required -- we know that our tunneling code needs ports. It was both easier and appropriate to rely on that existing logic to handle missing ports in that PR code.

In this low-level code, there is no existing framework to properly handle missing ports, no consensus that missing ports are actually a _problem_, and, after establishing that consensus, the missing code would need to be added in very different place(s) -- the problem cannot be properly addressed in _this_ method. That non-trivial out-of-scope change deserves a dedicated PR.